kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -201,7 +201,8 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
                         // try lock
                         val update = if (owner == null) EMPTY_LOCKED else Empty(owner)
                         if (_state.compareAndSet(state, update)) { // locked
-                            cont.resume(Unit)
+                            // TODO implement functional type in LockCont as soon as we get rid of legacy JS
```

Comment:
Btw, if that helps, we can optimize it for `onwer==null` case by caching a reference to a preallocated `onCancellationUnlock` reference similar to the way it is done in Semaphore.